File: sdks/elixir/lib/autonoma/refs.ex
defmodule Autonoma.Refs do
  @moduledoc """
  JWT-like signing primitive: `header.payload.signature` using HMAC-SHA256.

  The handler signs `%{"refs" => teardown, "testRunId" => ..., "environment" => name}`
  into the teardown token on `up` and verifies it back on `down`. The primitive
  itself is payload-agnostic and unchanged across the v2 migration.
  """

  @doc """
  Sign a refs payload into a 3-part token string.
  """
  def sign(payload, secret) do
    header = base64url_encode(Jason.encode!(%{"alg" => "HS256", "typ" => "REFS"}))
    # Sanitize before encoding so DateTime / NaiveDateTime / Decimal handles a
    # scenario may carry in `teardown` survive the JSON round-trip.
    body = base64url_encode(Jason.encode!(sanitize_for_json(payload)))
    signature = hmac_sign("#{header}.#{body}", secret)
    "#{header}.#{body}.#{signature}"
  end

  @doc """
  Verify and decode a teardown token. Returns the payload map or raises.
  """
  def verify!(token, secret) do
    parts = String.split(token, ".")

    if length(parts) != 3 do
      raise "malformed token"
    end

    [header, body, signature] = parts
    expected = hmac_sign("#{header}.#{body}", secret)

    if !Autonoma.HMAC.secure_compare(expected, signature) do
      raise "signature mismatch"
    end

    body
    |> base64url_decode()
    |> Jason.decode!()
  end

  @doc """
  Recursively sanitize values that Jason can't encode natively.
  Converts DateTime, NaiveDateTime, Date, Decimal, and other structs to strings.
  """
  def sanitize_for_json(%DateTime{} = dt), do: DateTime.to_iso8601(dt)
  def sanitize_for_json(%NaiveDateTime{} = dt), do: NaiveDateTime.to_iso8601(dt)
  def sanitize_for_json(%Date{} = d), do: Date.to_iso8601(d)
  def sanitize_for_json(%{__struct__: _} = struct) do
    # Handle Decimal and other structs by converting to string
    to_string(struct)
  end
  def sanitize_for_json(map) when is_map(map) do
    Map.new(map, fn {k, v} -> {k, sanitize_for_json(v)} end)
  end
  def sanitize_for_json(list) when is_list(list) do
    Enum.map(list, &sanitize_for_json/1)
  end
  def sanitize_for_json(tuple) when is_tuple(tuple) do
    tuple |> Tuple.to_list() |> Enum.map(&sanitize_for_json/1)
  end
  def sanitize_for_json(value), do: value

  defp base64url_encode(data) do
    Base.url_encode64(data, padding: false)
  end

  defp base64url_decode(data) do
    # Add padding back if needed
    padded =
      case rem(byte_size(data), 4) do
        2 -> data <> "=="
        3 -> data <> "="
        _ -> data
      end

    Base.url_decode64!(padded)
  end

  defp hmac_sign(data, secret) do
    :crypto.mac(:hmac, :sha256, secret, data)
    |> Base.url_encode64(padding: false)
  end
end

File: sdks/go/autonoma/refs.go
package autonoma

import (
	"crypto/hmac"
	"crypto/sha256"
	"crypto/subtle"
	"encoding/base64"
	"encoding/json"
	"errors"
	"strings"
)

// RefsPayload is the payload signed into the teardown token.
type RefsPayload struct {
	// Refs is whatever a scenario's Up returned as refs - arbitrary JSON, signed
	// at Up and handed back to Down verbatim.
	Refs any `json:"refs"`
	// TestRunID is the testRunId captured at Up time.
	TestRunID string `json:"testRunId"`
	// Environment is the scenario name. Named "environment" for wire/back-compat
	// reasons; Down reads it to route to the right scenario's teardown.
	Environment string `json:"environment"`
}

// SignRefs signs a refs payload into a JWT-like token (header.payload.signature)
// using HMAC-SHA256.
func SignRefs(payload RefsPayload, secret string) (string, error) {
	headerJSON, err := json.Marshal(map[string]string{"alg": "HS256", "typ": "REFS"})
	if err != nil {
		return "", err
	}
	payloadJSON, err := json.Marshal(payload)
	if err != nil {
		return "", err
	}

	header := base64.RawURLEncoding.EncodeToString(headerJSON)
	body := base64.RawURLEncoding.EncodeToString(payloadJSON)
	sig := hmacBase64URL(header+"."+body, secret)

	return header + "." + body + "." + sig, nil
}

// VerifyRefs verifies and decodes a teardown token, returning the payload or an error.
func VerifyRefs(token string, secret string) (*RefsPayload, error) {
	parts := strings.Split(token, ".")
	if len(parts) != 3 {
		return nil, errors.New("malformed token")
	}

	header, body, signature := parts[0], parts[1], parts[2]
	expected := hmacBase64URL(header+"."+body, secret)
	if subtle.ConstantTimeCompare([]byte(expected), []byte(signature)) != 1 {
		return nil, errors.New("signature mismatch")
	}

	payloadBytes, err := base64.RawURLEncoding.DecodeString(body)
	if err != nil {
		return nil, err
	}

	var payload RefsPayload
	if err := json.Unmarshal(payloadBytes, &payload); err != nil {
		return nil, err
	}

	return &payload, nil
}

func hmacBase64URL(data string, secret string) string {
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(data))
	return base64.RawURLEncoding.EncodeToString(mac.Sum(nil))
}
